stdlib/type.mli

```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
 
     At runtime, this function simply returns its second argument unchanged.
 *)
+
+(** {1:identifiers Type identifiers} *)
+
+(** Type identifiers.
+
+    Type identifiers are values that can be tested for
+    {{!Id.equal}equality} to prove they represent the same type. Note
+    that unequal identifiers does not imply unequal types: the same
+    type can be represented by multiple identifiers.
+
+    See an {{!Id.example}example} of use. *)
+module Id : sig
+
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
+  (** The type for type identifiers. ['a] is the represented type. *)
+
+  val make : unit -> 'a t
+  (** [make ()] is a new type identifier. *)
+
+  val equal : 'a t -> 'b t -> ('a, 'b) eq option
+  (** [equal id0 id1] is [Some Equal] if identifier [id0] is equal to [id1]
+      and [None] otherwise. *)
+
+  (** {1:example Example}
+
+      The following shows how type identifiers can be used to implement
```

Comment:
Sorry, I'm having trouble understanding this example, and maybe my sources of confusion can suggest some improvements:
- What is a "heterogenous dictionary" , when is it useful? This signature seems a bit useless to me right now - I still have to collect the keys somewhere to be able to look things up, and the keys are of different types... so do I need another heterogenous collection somewhere? (Possibly this is motivated by [hmap](https://erratique.ch/software/hmap/doc/Hmap.html), maybe exposing the `Key.hide_type` is essential for this to be useful?) 
- Typically dictionary "keys" are not opaque things you have to create/concoct (e.g. words -> definitions, you already know the words), is dictionary the right name to use for this?
- The `key` function is really confusing - you're creating something (`'a t`) out of nothing (`unit`) using a cyclical value that can be of any type you cast it to (because memory representations are equivalent?) - it's not obvious that the comparisons in `remov_assq` and `assq_opt` are halal for this cyclical value - at least not until you look at their documentation or implementation to see that they use physical equality and not structural. Is there perhaps a less nifty, more beginner-friendly version of this? Even the `Type.Id.make` implementation was easier to understand for me than this, magic-module syntax notwithstanding.

Reply:
An example of "heterogeneous dictionary" is a record, with a fixed number of fields (keys) that can access values of different types.

Reply:
> * What is a "heterogenous dictionary" , when is it useful?

Yes, I'll add a few words about that.

> Possibly this is motivated by [hmap](https://erratique.ch/software/hmap/doc/Hmap.html), maybe exposing the Key.hide_type is essential for this to be useful?)

No, in fact once you hide the type, it becomes less useful (unless you hide it in conjunction with something interesting like a function acting on the hidden type).  

> * Typically dictionary "keys" are not opaque things you have to create/concoct (e.g. words -> definitions, you already know the words), is dictionary the right name to use for this?

There are various names for structures that allow you to map key to values, hash tables, maps, [property lists](http://mlton.org/PropertyList), dictionaries, key-value stores. I just took a name that is not used by the stdlib and avoids fights about module casing conventions :–) If you want something else propose a better name. `Store` could be one but I think I rather keep it that way.

> * it's not obvious that the comparisons in `remov_assq` and `assq_opt` are halal for this cyclical value - at least not until you look at their documentation or implementation to see that they use physical equality and not structural

There's not much I can do about the name of these function which are likely inspired by lisp conventions. I notice though that, "the name is weird let's look up the docs" worked here :–)

Reply:
This structure reminded me of tickets as in "cloakroom ticket", so that could be another name for keys. This indicates they are created by the structure instead of chosen by the client and the name is already used in concurrency algorithms: ticket locks. I think "store" would be good.

If `'a key` is replaced with `'a Type.Id.t` similarly to my previous suggestion, it is simply an association list associating values of different types with their type identifiers.

Reply:
> There's not much I can do about the name of these function which are likely inspired by lisp conventions. I notice though that, "the name is weird let's look up the docs" worked here :–)  

It wasn't the names that had me confused - I had not encountered cyclical values before, and had to read [this answer](https://stackoverflow.com/a/44369585) first to understand a little more what was going on. Also, maybe some explanation around `exist_key` would be helpful? For example, why can't you just use `tid` to find things in the list, why do you need the `exist_key`?  

fwiw, I think it's also a fair position, to say that if you don't understand this part of the `Stdlib`, then don't worry about it - fine line to walk between api documentation and full-fledged blog post.

Also, do we need a note about not marshalling these? [This answer](https://discuss.ocaml.org/t/types-as-first-class-citizens-in-ocaml/2030/3) which references [this library](https://github.com/let-def/distwit) seems to suggest so.

Reply:
>  Also, maybe some explanation around exist_key would be helpful? For example, why can't you just use tid to find things in the list, why do you need the exist_key?

That's for the compiler to tell you :-) 

In any case if we add @hhugo's suggestion we can simplify the example and use regular assoc lists.

> fwiw, I think it's also a fair position, to say that if you don't understand this part of the Stdlib, then don't worry about it

My take is a bit different: here's a (possibly confusing) example, spend a bit of quality time fiddling with it, you could end up learning quite a bit from it – as you did.